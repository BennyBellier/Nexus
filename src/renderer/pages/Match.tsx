import { useState } from 'react';
import Main from '../components/Main';

const homeData = {
  name: 'Team1',
  score: 1,
  faults: 1,
  timeouts: 0,
};

const awayData = {
  name: 'Team2',
  score: 0,
  faults: 0,
  timeouts: 1500,
};

function Timer() {
  const [time, setTime] = useState(1599);
  // const [isRunning, setIsRunning] = useState(false);
  const [percentage, setPercentage] = useState(75);
  const [escaped, setEscaped] = useState(false);

  function tick() {
    setTime(time + 1);
    setPercentage(Math.ceil((time / 100) * 20));
    if (time >= 20000) {
      setEscaped(true);
    }
  }

  function timeToString() {
    const seconds = Math.floor(time / 100);
    const centiseconds = time % 100;

    const secondsString = seconds < 10 ? `0${seconds}` : `${seconds}`;
    const centisecondsString =
      centiseconds < 10 ? `0${centiseconds}` : `${centiseconds}`;

    return `${secondsString}:${centisecondsString}`;
  }

  return (
    <div className="rounded-2xl flex items-center justify-end overflow-hidden shadow-2xl">
      <div
        className={`w-full h-full flex flex-row justify-start items-center ${
          escaped ? 'before:bg-green-500' : 'before:bg-red-600'
        } before:h-full before:w-[${percentage.toString()}%]`}
      />
      <span className="border-l-2 border-slate-300 h-full text-center w-1/5 text-3xl flex justify-center items-center text-slate-800">
        {timeToString()}
      </span>
    </div>
  );
}

function Round({ maxRound, currentRound }: any) {
  return (
    <div className="bg-white rounded-2xl grid grid-rows-3 grid-cols-1 justify-center items-center text-3xl text-slate-800 shadow-2xl">
      <span className="text-center">
        {currentRound === undefined ? 0 : currentRound}
      </span>
      <hr className="w-3/5 text-slate-300 mx-auto" />
      <span className="text-center">
        {maxRound === undefined ? 0 : maxRound}
      </span>
    </div>
  );
}

function TeamElement({ team }: any) {
  function timeToString(time: number) {
    const seconds = Math.floor(time / 100);
    const centiseconds = time % 100;

    const secondsString = seconds < 10 ? `0${seconds}` : `${seconds}`;
    const centisecondsString =
      centiseconds < 10 ? `0${centiseconds}` : `${centiseconds}`;

    return `${secondsString}:${centisecondsString}`;
  }

  return (
    <div className="flex flex-col items-center gap-8 p-5 shadow-2xl bg-white rounded-2xl m-20 scale-150">
      <span className="text-3xl">{team.name}</span>
      <hr className="w-3/5 text-slate-300" />
      <div className="flex flex-row gap-8">
        <div className="flex flex-col gap-2 text-2xl text-center font-medium">
          <span className="text-slate-400">Score</span>
          <span>{team.score}</span>
        </div>
        <div className="flex flex-col gap-2 text-2xl text-center font-medium">
          <span className="text-slate-400">Fautes</span>
          <span>{team.faults}</span>
        </div>
      </div>
      <div className="flex flex-col gap-2 text-2xl text-center font-medium">
        <span
          className={`${
            timeToString(team.timeouts) !== '00:00'
              ? 'text-slate-800'
              : 'text-white'
          }`}
        >
          {timeToString(team.timeouts)}
        </span>
      </div>
    </div>
  );
}

function Teams({ home, away }: any) {
  return (
    <div className="w-full h-full flex flex-row justify-between items-center">
      <TeamElement team={home} />
      <TeamElement team={away} />
    </div>
  );
}

export default function Match() {
  const phase = 'Phase de poules';

  return (
    <Main title={phase} className="grid grid-rows-[auto_1fr_3fr]">
      <div className="grid grid-cols-[5fr_1fr] py-10 gap-5">
        <Timer />
        <Round />
      </div>
      <Teams home={homeData} away={awayData} />
    </Main>
  );
}
