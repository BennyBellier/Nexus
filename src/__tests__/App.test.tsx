import React from 'react';
import '@testing-library/jest-dom';
import { render, fireEvent } from '@testing-library/react';
import Match from '../renderer/pages/Match';

describe('Match component', () => {
  let component: any;
  let mockTimeUpdated: jest.Mock<any, any>;
  const mockMatch = {
    timeUpdated: jest.fn(),
    startStop: jest.fn(),
    scoreUpdated: jest.fn(),
    nextRound: jest.fn(),
    loaded: jest.fn(),
    init: jest.fn(),
  };
  window.match = mockMatch;
  const mockNexus = {
    getAssetsPath: jest.fn(),
    descriptionUpdated: jest.fn(),
    tcp_ip: jest.fn(),
    tcp_clientsUpdated: jest.fn(),
  };
  window.nexus = mockNexus;

  beforeEach(() => {
    component = render(<Match />);
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('renders the component', () => {
    window.match.timeUpdated(mockTimeUpdated, true);
    expect(component.container).toMatchSnapshot();
  });

  it('renders the create button', () => {
    const button = component.getByText('Nouveau match');
    expect(button).toBeInTheDocument();
  });

  it('renders the QR code button', () => {
    const button = component.getByText('Afficher le QR code');
    expect(button).toBeInTheDocument();
  });

  it('renders the correction button', () => {
    const checkbox = component.getByRole('switch');
    expect(checkbox).toBeInTheDocument();
  });

  it('clicks the start/stop button', () => {
    // Find the button by its ID
    const startStopButton = component.getByTestId('start-stop-timer');

    // Mock the startStop function
    const mockStartStop = jest.fn();
    window.match.startStop = mockStartStop;

    // Click the button
    fireEvent.click(startStopButton);

    // Check that the startStop function was called
    expect(mockStartStop).toHaveBeenCalled();
  });

  it('calls timeUpdated when the match time is updated', () => {
    mockTimeUpdated = jest.fn();
    expect(mockTimeUpdated).not.toHaveBeenCalled();
    // Simuler la réponse de la fonction timeUpdated
    window.match.timeUpdated((time: number, percentage: number) => {
      mockTimeUpdated(time, percentage);
    });

    // Ajouter un délai pour laisser le temps à la simulation de s'exécuter
    setTimeout(() => {
      // Vérifier que la fonction timeUpdated est appelée
      expect(mockTimeUpdated).toHaveBeenCalledWith(1000, 50);
    }, 100);
  });
});
