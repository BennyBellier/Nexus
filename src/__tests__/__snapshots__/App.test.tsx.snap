// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Match component renders the component 1`] = `
<div>
  <main
    class="bg-transparent overflow-hidden w-full p-4 text-slate-800 grid grid-rows-[auto_1fr_3fr]"
  >
    <div
      class="flex flex-row justify-between items-center max-h-9 h-9"
    >
      <div
        class="flex flex-row items-center gap-4"
      >
        <h1
          class="text-xl font-semibold"
        >
          Match
        </h1>
      </div>
      <div
        class="w-[50%] flex flex-row items-center justify-end gap-10 pr-5"
      >
        <button
          class="4h-fit px-4 py-1.5 rounded-lg border border-slate-300 hover:bg-slate-100 hover:scale-[99%] duration-200 hover:border-cyan-200 ease-in-out flex items-center gap-2"
          type="button"
        >
          <svg
            fill="none"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 15 15"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M7.49991 0.876892C3.84222 0.876892 0.877075 3.84204 0.877075 7.49972C0.877075 11.1574 3.84222 14.1226 7.49991 14.1226C11.1576 14.1226 14.1227 11.1574 14.1227 7.49972C14.1227 3.84204 11.1576 0.876892 7.49991 0.876892ZM1.82707 7.49972C1.82707 4.36671 4.36689 1.82689 7.49991 1.82689C10.6329 1.82689 13.1727 4.36671 13.1727 7.49972C13.1727 10.6327 10.6329 13.1726 7.49991 13.1726C4.36689 13.1726 1.82707 10.6327 1.82707 7.49972ZM7.50003 4C7.77617 4 8.00003 4.22386 8.00003 4.5V7H10.5C10.7762 7 11 7.22386 11 7.5C11 7.77614 10.7762 8 10.5 8H8.00003V10.5C8.00003 10.7761 7.77617 11 7.50003 11C7.22389 11 7.00003 10.7761 7.00003 10.5V8H4.50003C4.22389 8 4.00003 7.77614 4.00003 7.5C4.00003 7.22386 4.22389 7 4.50003 7H7.00003V4.5C7.00003 4.22386 7.22389 4 7.50003 4Z"
              fill="currentColor"
              fill-rule="evenodd"
            />
          </svg>
           Nouveau match
        </button>
        <div
          class="flex flex-row gap-2 justify-center items-center"
        >
          <input
            class="absolute appearance-none h-[34px] w-28 rounded-lg outline-none cursor-pointer border border-slate-300 hover:border-cyan-200 checked:bg-slate-100 checked:border-cyan-200"
            role="switch"
            type="checkbox"
          />
          <svg
            class="z-10 select-none pointer-events-none"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 24 24"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M0 0h24v24H0z"
              fill="none"
            />
            <path
              d="M20 7l.94-2.06L23 4l-2.06-.94L20 1l-.94 2.06L17 4l2.06.94zM8.5 7l.94-2.06L11.5 4l-2.06-.94L8.5 1l-.94 2.06L5.5 4l2.06.94zM20 12.5l-.94 2.06-2.06.94 2.06.94.94 2.06.94-2.06L23 15.5l-2.06-.94zM17.71 9.12l-2.83-2.83c-.2-.19-.45-.29-.71-.29-.26 0-.51.1-.71.29L2.29 17.46a.996.996 0 000 1.41l2.83 2.83c.2.2.45.3.71.3s.51-.1.71-.29l11.17-11.17c.39-.39.39-1.03 0-1.42zm-3.54-.7l1.41 1.41L14.41 11 13 9.59l1.17-1.17zM5.83 19.59l-1.41-1.41L11.59 11 13 12.41l-7.17 7.18z"
            />
          </svg>
          <span
            class="text-slate-800 z-10 select-none pointer-events-none"
          >
            Correction
          </span>
        </div>
        <button
          class="4h-fit px-4 py-1.5 rounded-lg border border-slate-300 hover:bg-slate-100 hover:scale-[99%] duration-200 hover:border-cyan-200 ease-in-out flex items-center gap-2"
          type="button"
        >
          <svg
            aria-hidden="true"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 20 20"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M3 4a1 1 0 011-1h3a1 1 0 011 1v3a1 1 0 01-1 1H4a1 1 0 01-1-1V4zm2 2V5h1v1H5zM3 13a1 1 0 011-1h3a1 1 0 011 1v3a1 1 0 01-1 1H4a1 1 0 01-1-1v-3zm2 2v-1h1v1H5zM13 3a1 1 0 00-1 1v3a1 1 0 001 1h3a1 1 0 001-1V4a1 1 0 00-1-1h-3zm1 2v1h1V5h-1z"
              fill-rule="evenodd"
            />
            <path
              d="M11 4a1 1 0 10-2 0v1a1 1 0 002 0V4zM10 7a1 1 0 011 1v1h2a1 1 0 110 2h-3a1 1 0 01-1-1V8a1 1 0 011-1zM16 9a1 1 0 100 2 1 1 0 000-2zM9 13a1 1 0 011-1h1a1 1 0 110 2v2a1 1 0 11-2 0v-3zM7 11a1 1 0 100-2H4a1 1 0 100 2h3zM17 13a1 1 0 01-1 1h-2a1 1 0 110-2h2a1 1 0 011 1zM16 17a1 1 0 100-2h-3a1 1 0 100 2h3z"
            />
          </svg>
           Afficher le QR code
        </button>
      </div>
    </div>
    <div
      class="grid grid-cols-[6fr_1fr] py-10 gap-10"
    >
      <button
        class="rounded-2xl flex items-center justify-end overflow-hidden shadow-2xl select-none"
        data-testid="start-stop-timer"
        id="StartStopTimer"
        name="StartStopTimer"
        type="button"
      >
        <div
          class="w-full h-full flex flex-row justify-start items-center overflow-hidden"
        >
          <div
            class="h-full bg-red-600"
            style="width: 0.00%;"
          />
        </div>
        <span
          class="border-l-2 border-slate-300 h-full text-center w-1/5 text-3xl flex justify-center items-center text-slate-800"
        >
          00:00
        </span>
      </button>
      <button
        class="bg-white rounded-2xl flex flex-col justify-around items-center text-3xl text-slate-800 shadow-2xl select-none"
        type="button"
      >
        <span
          class="text-center select-none"
        >
          0
        </span>
        <hr
          class="w-3/5 text-slate-300 mx-auto"
        />
        <span
          class="text-center select-none"
        >
          12
        </span>
      </button>
    </div>
    <div
      class="w-full h-full flex flex-row justify-around items-center"
    >
      <div
        class="  flex flex-col items-center gap-8 p-5 shadow-2xl bg-white rounded-2xl scale-150 select-none"
      >
        <span
          class="text-3xl"
        />
        <hr
          class="w-3/5 text-slate-300"
        />
        <div
          class="flex flex-row gap-8"
        >
          <div
            class="flex flex-col gap-2 text-2xl text-center font-medium"
          >
            <span
              class="text-slate-400"
            >
              Score
            </span>
            <span />
          </div>
          <div
            class="flex flex-col gap-2 text-2xl text-center font-medium"
          >
            <span
              class="text-slate-400"
            >
              Fautes
            </span>
            <span />
          </div>
        </div>
        <div
          class="flex flex-col gap-2 text-2xl text-center font-medium"
        >
          <span
            class="text-white select-none"
          >
            00:00
          </span>
        </div>
      </div>
      <div
        class=" shadow-cyan-200 flex flex-col items-center gap-8 p-5 shadow-2xl bg-white rounded-2xl scale-150 select-none"
      >
        <span
          class="text-3xl"
        />
        <hr
          class="w-3/5 text-slate-300"
        />
        <div
          class="flex flex-row gap-8"
        >
          <div
            class="flex flex-col gap-2 text-2xl text-center font-medium"
          >
            <span
              class="text-slate-400"
            >
              Score
            </span>
            <span />
          </div>
          <div
            class="flex flex-col gap-2 text-2xl text-center font-medium"
          >
            <span
              class="text-slate-400"
            >
              Fautes
            </span>
            <span />
          </div>
        </div>
        <div
          class="flex flex-col gap-2 text-2xl text-center font-medium"
        >
          <span
            class="text-white select-none"
          >
            00:00
          </span>
        </div>
      </div>
    </div>
    <div
      class=" hidden absolute z-{999999} w-screen h-screen flex justify-center items-center bg-opacity-40 bg-gray-600 -translate-x-4 -translate-y-4 duration-500"
    >
      <div
        class="absolute bg-white flex justify-center items-center left-1/2 top-1/2 h-1/2 w-fit rounded-xl shadow-3xl -translate-x-3/4 -translate-y-1/2 p-10 pt-11"
      >
        <button
          class="p-1 text-white bg-orange-500 rounded-full absolute top-4 right-0 -translate-x-2 -translate-y-2"
          type="button"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            viewBox="0 0 16 16"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              clip-rule="evenodd"
              d="M8 8.707l3.646 3.647.708-.707L8.707 8l3.647-3.646-.707-.708L8 7.293 4.354 3.646l-.707.708L7.293 8l-3.646 3.646.707.708L8 8.707z"
              fill-rule="evenodd"
            />
          </svg>
        </button>
        <svg
          class="w-full h-full"
          height="128"
          viewBox="0 0 21 21"
          width="128"
        >
          <path
            d="M0,0 h21v21H0z"
            fill="#FFFFFF"
            shape-rendering="crispEdges"
          />
          <path
            d="M0 0h7v1H0zM9 0h1v1H9zM14,0 h7v1H14zM0 1h1v1H0zM6 1h1v1H6zM10 1h2v1H10zM14 1h1v1H14zM20,1 h1v1H20zM0 2h1v1H0zM2 2h3v1H2zM6 2h1v1H6zM8 2h5v1H8zM14 2h1v1H14zM16 2h3v1H16zM20,2 h1v1H20zM0 3h1v1H0zM2 3h3v1H2zM6 3h1v1H6zM8 3h3v1H8zM12 3h1v1H12zM14 3h1v1H14zM16 3h3v1H16zM20,3 h1v1H20zM0 4h1v1H0zM2 4h3v1H2zM6 4h1v1H6zM9 4h3v1H9zM14 4h1v1H14zM16 4h3v1H16zM20,4 h1v1H20zM0 5h1v1H0zM6 5h1v1H6zM9 5h1v1H9zM11 5h2v1H11zM14 5h1v1H14zM20,5 h1v1H20zM0 6h7v1H0zM8 6h1v1H8zM10 6h1v1H10zM12 6h1v1H12zM14,6 h7v1H14zM3 8h2v1H3zM6 8h2v1H6zM12 8h1v1H12zM17 8h2v1H17zM0 9h2v1H0zM5 9h1v1H5zM7 9h2v1H7zM11 9h2v1H11zM15 9h3v1H15zM19,9 h2v1H19zM0 10h2v1H0zM5 10h4v1H5zM12 10h4v1H12zM17 10h1v1H17zM20,10 h1v1H20zM1 11h1v1H1zM3 11h2v1H3zM8 11h4v1H8zM15 11h2v1H15zM19 11h1v1H19zM0 12h1v1H0zM2 12h2v1H2zM6 12h1v1H6zM8 12h1v1H8zM14,12 h7v1H14zM8 13h2v1H8zM13 13h1v1H13zM18,13 h3v1H18zM0 14h7v1H0zM8 14h3v1H8zM13 14h2v1H13zM17 14h2v1H17zM20,14 h1v1H20zM0 15h1v1H0zM6 15h1v1H6zM9 15h3v1H9zM13 15h2v1H13zM18 15h1v1H18zM0 16h1v1H0zM2 16h3v1H2zM6 16h1v1H6zM8 16h2v1H8zM11 16h1v1H11zM14 16h1v1H14zM16 16h1v1H16zM18 16h2v1H18zM0 17h1v1H0zM2 17h3v1H2zM6 17h1v1H6zM8 17h2v1H8zM12 17h1v1H12zM15 17h2v1H15zM0 18h1v1H0zM2 18h3v1H2zM6 18h1v1H6zM9 18h1v1H9zM11 18h2v1H11zM15 18h3v1H15zM19,18 h2v1H19zM0 19h1v1H0zM6 19h1v1H6zM13 19h1v1H13zM15 19h1v1H15zM17 19h1v1H17zM19,19 h2v1H19zM0 20h7v1H0zM10 20h3v1H10zM14 20h3v1H14zM18 20h2v1H18z"
            fill="#000000"
            shape-rendering="crispEdges"
          />
        </svg>
      </div>
    </div>
    <div
      class="absolute bottom-2 right-2 flex flex-row gap-1 justify-center items-center"
    >
      0
      <span
        class=" text-red-500 -translate-y-[1px]"
      >
        <svg
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 12 12"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <circle
            cx="6"
            cy="6"
            fill-rule="evenodd"
            r="5"
            stroke="#000"
          />
        </svg>
      </span>
    </div>
  </main>
</div>
`;
